google_authorized_keys/main.go

```diff
@@ -0,0 +1,178 @@
+//  Copyright 2022 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+// GoogleAuthorizedKeys obtains SSH keys from metadata.
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"runtime"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/GoogleCloudPlatform/guest-agent/utils"
+	"github.com/GoogleCloudPlatform/guest-logging-go/logger"
+	"github.com/tarm/serial"
+)
+
+var (
+	programName       = "GoogleAuthorizedKeysCommand"
+	version           = "1.0.0"
+	metadataURL       = "http://169.254.169.254/computeMetadata/v1/"
+	metadataRecursive = "/?recursive=true&alt=json"
+	metadataHang      = "&wait_for_change=true&timeout_sec=2"
+	defaultTimeout    = 2 * time.Second
+)
+
+func logFormat(e logger.LogEntry) string {
+	now := time.Now().Format("2006/01/02 15:04:05")
+	return fmt.Sprintf("%s %s: %s", now, programName, e.Message)
+}
+
+type serialPort struct {
+	port string
+}
+
+func (s *serialPort) Write(b []byte) (int, error) {
+	c := &serial.Config{Name: s.port, Baud: 115200}
+	p, err := serial.OpenPort(c)
+	if err != nil {
+		return 0, err
+	}
+	defer p.Close()
+
+	return p.Write(b)
+}
+
+func getMetadataKey(key string) (string, error) {
+	md, err := getMetadata(key, false)
+	if err != nil {
+		return "", err
+	}
+	return string(md), nil
+}
+
+func getMetadata(key string, recurse bool) ([]byte, error) {
```

Comment:
I think that the serialPort stuff can. I've removed getMetadataKey because it was no longer needed, and getMetadata is different. I think making the serialPort stuff shared might be better as a separate PR though.